### ISSUE TITLE
Increasing latency as # of connections increase. 

### ISSUE THREAD

**Issue by michealp-coder** [mtrudel/bandit#72]
### TL;DR;

I wanted to raise awareness to the non-linear increase in latency/TTFB as # of Connections increases (especially as # of connections >= 256). This isn’t typical of Erlang/OTP and might warrant further investigation. 

### Context:

A major characteristic of Erlang/OTP is to respond fast or not respond at all. Other languages might have higher RPS, but Erlang has the most consistent performance as measured in response time (which is flat and extremely low TTFB under load). 

This is a major reason why people choose Erlang because they know under load, customer experience won’t be impacted since response time will always be fast. 

### Benchmarks:

You can see that desirable scaling curve play out in the below language benchmarks.

Example A, notice Cowboy 1.x consistently low/flat response times (red solid line):

![](https://stressgrid.com/blog/webserver_benchmark/response_lat90.png)

Example B, notice Cowboy 1.x low response times (red solid line) vs 2.x:

![](https://stressgrid.com/blog/cowboy_performance/response_lat99.png)

The accompanying blog posts, where the graphs can be found:

- https://stressgrid.com/blog/webserver_benchmark/
- https://stressgrid.com/blog/cowboy_performance/

### Observation:

In the Bandit vs Cowboy benchmark, Bandits displays ever increasing latency (where Cowboy TTFB has a smaller variance) as # of a connections increase. 

Essentially, Bandit is not scaling like typical Erlang/OTP, where it has consistently low TTFB under load. 

- On h2c / 16 streams, **Bandit has upwards of 30x longer TTFB compared to Cowboy** (Bandits taking multiple seconds to respond relative to Cowboy). 

- Note: even when Cowboy has errors, it’s still responding fast (low TTFB). 

https://github.com/mtrudel/network_benchmark/blob/0b18a9b299b9619c38d2a70ab967831565121d65/benchmarks-09-2021.pdf

### Ask:

Given that the 0.6x series is to work on performance - it might warrant investigation into how to achieve a more consistent low latency (fast TTFB) as # of Connections increase. Essentially, scale with regards to TTFB more similar to Cowboy 1.x, where TTFB remains consistently low. Hopefully this doesn’t come at the cost of RPS, which makes Bandit so great. 

_Please don’t take my comments as being negative. You’ve created a phenomenal web server, and want to express appreciation. So thank you in advance for all your great work on Bandit._

**Comment by ryanwinchester** [comment]
> * https://stressgrid.com/blog/cowboy_performance/

The part 2 was especially good read

**Comment by mtrudel** [comment]
I'd noticed this when I was doing the original benchmarking you link to above and, truthfully, forgot about it. I agree it's a critical thing to address! In terms of a concrete plan to do so:

1. The first step here is to get up-to-date objective & reproducible numbers that demonstrate the issue. The benchmarks you quote above at mtrudel/network_benchmarks reflect a one-off benchmark I ran against Bandit 0.4.x; a LOT has changed since then (though I can't think of any changes which addresses TTFB latency specifically).

2. 0.6.x adds a systematized version of the benchmark above to CI. Taking a look at the [CSV export of a recent benchmark run (https://github.com/mtrudel/bandit/actions/runs/3659270426), I can't see any evidence of time to request numbers that aren't broadly correlated between Bandit and Cowboy (though there are a LOT of missing numbers). Getting trustable numbers here is the first thing to be done.

3. I'll figure out why there are so many missing numbers in CI, and add a set of graphs to compare TTFB results as well as RPS results on every run. 

4. From there we can assess if the disparity is still present, and work on mitigation for it.

**Comment by mtrudel** [comment]
In terms of timing, I'm elbow deep in perf improvements in the WebSocket stack at the moment, and will dig into this after I'm done there. Likely in the first few weeks of 2023.

**Comment by michealp-coder** [comment]
@mtrudel That's awesome!

I do want to point out, like @ryanwinchester mentions, that Part 2 of the blog post was super informative. 

Essentially, Cowboy 1.x had better performance curves (than 2.x) and it was only after this blog was posted that Cowboy 2.x addressed it; however, I've never seen updated benchmarks.

Again, many thanks for the awesome software you've created.

**Comment by mtrudel** [comment]
Those two articles heavily influenced my approach to Thousand Island design, specifically around supporting HTTP/2.

In a nutshell, Bandit's process model is the best of both worlds; for HTTP/1.1 connections each client connection (possibly consisting of multiple keepalive requests) is handled in a single connection process (the Cowboy 1.x model). HTTP/2 connections use a split process approach closer to the Cowboy 2.x model, where the connection process is responsible for all 'client-facing' aspects of the connection, and each steam/request is handled by its own spawned sub-process (these processes are spawned / linked directly from the connection process in lieu of a big-S Supervisor model, since failure handling is pretty domain specific).

**Comment by mtrudel** [comment]
An update here: I figured out why we're not getting any TTFB / TTC / TTR request numbers in the benchmark CSV; it was [pretty odd](https://github.com/mtrudel/benchmark/commit/5ee4b306c9a88ff6a6cedd6a87f5ca79bb0dd929). 

I ran a [comparative benchmark](https://github.com/mtrudel/bandit/actions/runs/3950921199) against Cowboy, and noted that while Bandit's TTFB and TTC numbers for HTTP/1.1 are quite a bit worse than Cowboy's, our TTR numbers are quite a bit better (see the definition of these terms [here](https://nghttp2.org/documentation/h2load.1.html#h2load-1-output)). This is telling me that while Bandit is slower to connect initially, it gets the requests completed sooner. 

I dug into things a bit and noticed that Cowboy uses 100 acceptor processes by default, while Bandit uses 10. Thinking that this may explain the discrepancy on TTFB, I've cut a branch that ups Bandit's acceptors to 100 and am [rerunning the benchmark](https://github.com/mtrudel/bandit/actions/runs/3951248918). Hopefully this mitigates the issue.

I'll report back here once the new benchmark completes and I've had a chance to look at the new numbers.

**Comment by mtrudel** [comment]
Alright. So I have good news and bad news (and then good news again):

First, the (first) good news: Bandit's Time to Request values (the time taken to *complete* a request) are consistently about 2-3x faster than Cowboy:

![](https://camo.githubusercontent.com/888c8521caf66f2782f1ea109dca63dcb42000f5c3357575214cbe81b0d7e727/68747470733a2f2f717569636b63686172742e696f2f63686172743f77696474683d353030266865696768743d33303026633d253742253232646174612532323a25374225323264617461736574732532323a253542253742253232646174612532323a2535422d34372e3637343431383630343635313136362c2d36382e393635353137323431333739332c2d34392e3338323731363034393338323731352c2d34372e38313837393139343633303837332c2d35302e302535442c25323266696c6c2532323a2532326f726967696e2532322c2532326c6162656c2532323a253232646f776e6c6f61642532322537442c253742253232646174612532323a2535422d36372e38353731343238353731343238362c2d38312e34353136313239303332323538312c2d37332e35323934313137363437303538382c2d36362e36313431373332323833343634362c2d36332e3131343735343039383336303635362535442c25323266696c6c2532323a2532326f726967696e2532322c2532326c6162656c2532323a2532326563686f2532322537442c253742253232646174612532323a2535422d34332e39303234333930323433393032352c2d36362e303337373335383439303536362c2d34352e3830363435313631323930333232342c2d34302e33323834363731353332383436372c2d33372e3631303631393436393032363534352535442c25323266696c6c2532323a2532326f726967696e2532322c2532326c6162656c2532323a2532326e6f6f702532322537442c253742253232646174612532323a2535422d36382e30373232383931353636323635312c2d38302e39313238363330373035333934322c2d37332e30323435323331363037363239342c2d36372e32382c2d36332e3132352535442c25323266696c6c2532323a2532326f726967696e2532322c2532326c6162656c2532323a25323275706c6f61642532322537442535442c2532326c6162656c732532323a253542312c322c342c31362c36342535442537442c2532326f7074696f6e732532323a2537422532327363616c65732532323a25374225323278417865732532323a2535422537422532327363616c654c6162656c2532323a253742253232646973706c61792532323a747275652c2532326c6162656c537472696e672532323a2532324e756d6265722532306f66253230436c69656e74732532322537442537442535442c25323279417865732532323a253542253742253232677269644c696e65732532323a2537422532327a65726f4c696e65436f6c6f722532323a253232677261792532322c2532327a65726f4c696e6557696474682532323a332537442c2532327363616c654c6162656c2532323a253742253232646973706c61792532323a747275652c2532326c6162656c537472696e672532323a25323250657263656e74253230446966666572656e63652532322537442537442535442537442c2532327469746c652532323a253742253232646973706c61792532323a747275652c253232746578742532323a25354225323262616e646974406d61696e25323076732e253230636f77626f79406d61737465722532322c253232687474702f312e312532304025323031253230636f6e63757272656e63792c25323074696d655f746f5f726571756573745f6d65616e2532322535442537442537442c253232747970652532323a2532326c696e65253232253744)

This tracks with the values we see for overall RPS, which makes sense (completing requests faster means that we can complete more of them in a given period of time).

Now the bad news: Bandit's Time to Connect / Time to First Byte values are still worse than Cowboy's: 

![](https://camo.githubusercontent.com/b1b23be8700b826c905135c91a8968b4c372a1d32931dfbdf663ba105839dcc8/68747470733a2f2f717569636b63686172742e696f2f63686172743f77696474683d353030266865696768743d33303026633d253742253232646174612532323a25374225323264617461736574732532323a253542253742253232646174612532323a253542313036342e302c31302e3230343038313633323635333035372c2d392e3039303930393039303930393039332c302e302c313034322e333732383831333535393332312535442c25323266696c6c2532323a2532326f726967696e2532322c2532326c6162656c2532323a253232646f776e6c6f61642532322537442c253742253232646174612532323a2535422d332e37352c31322e3335393535303536313739373734372c35332e3434383237353836323036383935342c37342e34313836303436353131363237382c3230382e36393536353231373339313330362535442c25323266696c6c2532323a2532326f726967696e2532322c2532326c6162656c2532323a2532326563686f2532322537442c253742253232646174612532323a25354231332e3739333130333434383237353835382c33342e31343633343134363334313436342c3431332e373933313033343438323735382c3131312e35333834363135333834363135352c2d32322e36343135303934333339363232372535442c25323266696c6c2532323a2532326f726967696e2532322c2532326c6162656c2532323a2532326e6f6f702532322537442c253742253232646174612532323a253542372e3031373534333835393634393132332c312e373835373134323835373134323931382c3332352e343233373238383133353539332c32322e32323232323232323232323232332c313936352e3930393039303930393039312535442c25323266696c6c2532323a2532326f726967696e2532322c2532326c6162656c2532323a25323275706c6f61642532322537442535442c2532326c6162656c732532323a253542312c322c342c31362c36342535442537442c2532326f7074696f6e732532323a2537422532327363616c65732532323a25374225323278417865732532323a2535422537422532327363616c654c6162656c2532323a253742253232646973706c61792532323a747275652c2532326c6162656c537472696e672532323a2532324e756d6265722532306f66253230436c69656e74732532322537442537442535442c25323279417865732532323a253542253742253232677269644c696e65732532323a2537422532327a65726f4c696e65436f6c6f722532323a253232677261792532322c2532327a65726f4c696e6557696474682532323a332537442c2532327363616c654c6162656c2532323a253742253232646973706c61792532323a747275652c2532326c6162656c537472696e672532323a25323250657263656e74253230446966666572656e63652532322537442537442535442537442c2532327469746c652532323a253742253232646973706c61792532323a747275652c253232746578742532323a25354225323262616e646974406d61696e25323076732e253230636f77626f79406d61737465722532322c253232687474702f312e312532304025323031253230636f6e63757272656e63792c25323074696d655f746f5f636f6e6e6563745f6d65616e2532322535442537442537442c253232747970652532323a2532326c696e65253232253744)

![](https://camo.githubusercontent.com/0c457a499c6af04ff7086ceac29c90180eeedeb2c786c47f970ced9d66e75986/68747470733a2f2f717569636b63686172742e696f2f63686172743f77696474683d353030266865696768743d33303026633d253742253232646174612532323a25374225323264617461736574732532323a253542253742253232646174612532323a2535423333352e333434383237353836323036392c31362e3631353338343631353338343631332c35312e38353939353632333633323338352c3139312e39373038303239313937303830342c3435372e373733313039323433363937352535442c25323266696c6c2532323a2532326f726967696e2532322c2532326c6162656c2532323a253232646f776e6c6f61642532322537442c253742253232646174612532323a253542392e3138363335313730363033363734362c2d32352e37353735373537353735373537352c31332e3734373232383338313337343732352c3134312e32383434303336363937323437382c36352e39373933383134343332393839362535442c25323266696c6c2532323a2532326f726967696e2532322c2532326c6162656c2532323a2532326563686f2532322537442c253742253232646174612532323a253542302c302c302c302c302535442c25323266696c6c2532323a2532326f726967696e2532322c2532326c6162656c2532323a2532326e6f6f702532322537442c253742253232646174612532323a253542302c302c302c302c302535442c25323266696c6c2532323a2532326f726967696e2532322c2532326c6162656c2532323a25323275706c6f61642532322537442535442c2532326c6162656c732532323a253542312c322c342c31362c36342535442537442c2532326f7074696f6e732532323a2537422532327363616c65732532323a25374225323278417865732532323a2535422537422532327363616c654c6162656c2532323a253742253232646973706c61792532323a747275652c2532326c6162656c537472696e672532323a2532324e756d6265722532306f66253230436c69656e74732532322537442537442535442c25323279417865732532323a253542253742253232677269644c696e65732532323a2537422532327a65726f4c696e65436f6c6f722532323a253232677261792532322c2532327a65726f4c696e6557696474682532323a332537442c2532327363616c654c6162656c2532323a253742253232646973706c61792532323a747275652c2532326c6162656c537472696e672532323a25323250657263656e74253230446966666572656e63652532322537442537442535442537442c2532327469746c652532323a253742253232646973706c61792532323a747275652c253232746578742532323a25354225323262616e646974406d61696e25323076732e253230636f77626f79406d61737465722532322c253232687474702f312e312532304025323031253230636f6e63757272656e63792c25323074696d655f746f5f66697273745f627974655f6d65616e2532322535442537442537442c253232747970652532323a2532326c696e65253232253744)

So even though we're faster *overall* in serving requests, we don't get to the *first* part of a request as fast as Cowboy does. What's more, I can't see a good explanation for this. I spent a bunch of time digging through Ranch/Cowboy's code to see if they had any tricks up their sleeve, and I can't find anything. We're doing almost exactly the same thing in exactly the same way, calling the same underlying APIs with the same set of options. So I'm at a bit of a loss as to how to improve things (I've got a small PR #82 to get a bit of a win, but it's not much).

Now, the (second) good news: I don't actually think any of this really matters too much. A couple of points:

* The chart at the very top of this PR reflects a totally different test setup than anything that I've ever tested against Bandit. I fully expect that Bandit's numbers would look nearly the same (or better) than Cowboy's were it to be put through that same test suite. Comparing those charts to anything in Bandit's benchmarking setup (including the links that @michealp-coder posted to the `network_benchmark` repo) is very much apples to oranges.

* In the large, I believe that TTR is a more important timing benchmark than TTFB. The fact that Bandit's TTR numbers are as good as they are says all that *most* people need to know about this, I think.

* The places that Bandit fares the worst with TTFB and TTC are way at the high end of concurrency (specifically, they're *way* beyond the number of cores on any given test server). To the extent that this may be a real-world problem, it will only be so in extreme conditions

All of this is to say that I don't think I'm going to spend a whole lot more time on this at the moment. I've added charts to CI benchmarking going forward, so we'll be able to identify any regressions in this regard. In the meantime, if anyone has any ideas on this, I'm all ears.

**Comment by michealp-coder** [comment]
@mtrudel any chance we can increase the # of clients in the benchmark from 64, to something much larger than 256?

I ask because one of the other variances in performance, and you can begin to see it in the new benchmarks, is that the performance differences between Cowboy & Bandit deminishes/shrinks as the # of clients increase above >256.

**Comment by mtrudel** [comment]
> @mtrudel any chance we can increase the # of clients in the benchmark from 64, to something much larger than 256?
> 
> I ask because one of the other variances in performance, and you can begin to see it in the new benchmarks, is that the performance differences between Cowboy & Bandit deminishes/shrinks as the # of clients increase above >256.

I [updated](https://github.com/mtrudel/benchmark/commit/923bb32377d23115edc09f641120839597706157) the size of the concurrency requests to range up to 256; all benchmark runs will now use these sizes (you can see such a cowboy benchmark [here](https://github.com/mtrudel/bandit/actions/runs/3959875046). 

I don't think going much beyond 256 is worth it on GitHub's (fairly small) CI workers. When / if they get around to larger VMs on public projects I'll most certainly be making use of those for this. 

In any case, patterns seem to be *much* more obvious with the new concurrencies; the decay pattern in the above comparative benchmark, for example, is very apparent.

I'll also note that the benchmark repo is easy to run locally / in the cloud if you're so inclined - it's totally self-contained and easy to use in this regard. See [the README](https://github.com/mtrudel/benchmark) for more!

**Comment by mtrudel** [comment]
In terms of the overall pattern of comparative benchmarks, it does / has always seemed to be the case that Bandit's advantage peaks at around the core count of the underlying VM, with diminishing returns as concurrency gets higher.

**Comment by michealp-coder** [comment]
@mtrudel one last thing, and again thank you so much for all you do, is there a reason why your ignored h2c in your analysis? Because it appears from the benchmarks that bandit & cowboy are very similar in overall rps at higher client counts. 

I have to imagine most people would be using bandit in h2c config, so it might be disingenuous to call bandit 2-3x faster than cowboy if that’s not how most would use bandit.

**Comment by mtrudel** [comment]
> @mtrudel one last thing, and again thank you so much for all you do, is there a reason why your ignored h2c in your analysis? Because it appears from the benchmarks that bandit & cowboy are very similar in overall rps at higher client counts.

I assume you're asking about ignoring h2c in the context of the work further up in this thread

> I have to imagine most people would be using bandit in h2c config, so it might be disingenuous to call bandit 2-3x faster than cowboy if that’s not how most would use bandit.

A couple of things here:

* I've never claimed that Bandit is any amount faster than Cowboy *in all cases*. The wording that I've **always** used publicly (as seen on the project README and in talks such as https://youtu.be/ZLjWyanLHuk?t=1320 and https://youtu.be/FtZBTUvRt0g?t=518) is 'up to *x* times faster in some scenarios', and have always gone out of my way to enumerate the differences between HTTP/1.1 and h2 performance. Moreover, I've put the relevant graphs and links to the benchmarks as front and centre as I can whenever I mention performance numbers. I fail to see anything disingenuous about this.

* In my experience, the majority of Elixir deployments are run behind a reverse proxy such as nginx or any of innumerable cloud platforms; almost all of these platforms exclusively use HTTP/1.1 as a means of connecting to an upstream server such as Phoenix. Moreover, the majority of WebSocket requests in the wild are sourced from HTTP/1.1 connections (the only browser to support RFC 8441 Chrome 91+, to my knowledge). I've correspondingly put most of my optimization energy into HTTP/1.1 as a stack, as I believe this is actually where the most value is to be found (if you've come across them, I'd love to see comparative protocol stats in the Elixir community though!).
